module/zfs/dnode.c

```diff
@@ -1815,6 +1880,8 @@ dnode_free_range(dnode_t *dn, uint64_t off, uint64_t len, dmu_tx_t *tx)
 		if (last != first)
 			dnode_dirty_l1(dn, last, tx);
 
+		dnode_dirty_l1range(dn, first, last, tx);
+
```

Comment:
dnode_dirty_l1range dirties all incore dbufs.  The loop below dirties all dbufs that have been written to disk.  If you have some dbufs that were just created, they wouldn't be dirtied by the loop below.